packages/plugin/src/Bundles/Form/Limiting/FormLimiting.php

```diff
@@ -141,12 +149,25 @@ private function limitLoggedInOnly(Form $form)
 
     private function addMessage(Form $form, $message = "Sorry, you've already submitted this form.")
     {
-        if (\in_array($form->getId(), $this->formCache, true)) {
-            return;
+        // Triggered during from validation
+        if ($this->event instanceof ValidationEvent) {
+            if (\in_array($form->getId(), $this->formCache, true)) {
+                return;
+            }
+
+            $form->addError(Freeform::t($message));
+
+            $this->formCache[] = $form->getId();
         }
 
-        $form->addError(Freeform::t($message));
+        // Triggered when form is loaded
+        if ($this->event instanceof FormLoadedEvent) {
+            $form->setSubmissionLimitReached(true);
+        }
```

Comment:
Great cleanup BTW, but just wanted to double check your suggestion and that I follow correctly.

Reply:
> Disregard the suggestion I made previously it was wrong, but please refactor it to have a local $event parameter that's passed to the addMessage function instead of having a class variable.

Okay